### PR TITLE
Core docs improvements

### PR DESCRIPTION
**Pull request by okuchaiev**
Signed-off-by: Oleksii Kuchaiev <okuchaiev@nvidia.com>

### CODE REVIEW

File: docs/source/core/core.rst [hunk 1]
```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
...
+
+        # This is true because VoidType is compatible with every other element type (SpectrogramType in this case)
+        # And axes layout between btc_void and btc_spctr is the same
+        assert btc_void.compare(btc_spctr) == NeuralTypeComparisonResult.SAME
+        # These two types are incompatible because even though VoidType is used for elements on one side,
+        # the axes layout is different
+        assert btc_void.compare(btc_spct_bad) == NeuralTypeComparisonResult.INCOMPATIBLE
+        # Note that even though VoidType is compatible with every other type, other types are not compatible with VoidType!
+        # It is one-way compatibility
+        assert btc_spctr.compare(btc_void) == NeuralTypeComparisonResult.INCOMPATIBLE
```
Comment (titu1994) [inline docs/source/core/core.rst:789]: Possibly in the future we might want to make it symmetric. Not for RC though,

File: docs/source/core/core.rst [hunk 2]
```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
...
+Enforcing dimensions
+~~~~~~~~~~~~~~~~~~~~
+In addition to specifying tensor layout and elements' semantics, neural types also allow you to enforce tensor dimensions.
+You will have to use long notation to specify dimensions. Short notation only allows you to specify axes semantics and assumes
+arbitrary dimensions.
+
+.. code-block:: python
+
+        type1 = NeuralType(
+        (AxisType(AxisKind.Batch, 64), AxisType(AxisKind.Time, 10), AxisType(AxisKind.Dimension, 128)),
```
Comment (titu1994) [inline docs/source/core/core.rst:837]: Do we enforce this inside .compare()? I know `typecheck()` does not do actual shape enforcement. Ill add to my todos if .compare() doesnt do it currently.

File: docs/source/core/core.rst [hunk 3]
```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
...
+==============
+NeMo is built around Neural Modules, conceptual blocks of neural networks that take typed inputs
+and produce typed outputs. Such modules typically represent data layers, encoders, decoders, language models, loss functions, or methods of combining activations.
+NeMo makes it easy to combine and re-use these building blocks while providing a level of semantic correctness checking via its neural type system.
 
-Neural Module
--------------
-Neural Modules are building blocks for Models.
-They accept (typed) inputs and return (typed) outputs. *All Neural Modules inherit from ``torch.nn.Module`` and, therefore, compatible with PyTorch ecosystem.* There are 3 types on Neural Modules:
+.. note:: *All Neural Modules inherit from ``torch.nn.Module`` and, therefore, compatible with PyTorch ecosystem.*
```
Comment (ericharper) [inline docs/source/core/core.rst:640]:
```suggestion
.. note:: *All Neural Modules inherit from ``torch.nn.Module`` and are therefore compatible with the PyTorch ecosystem.*
```

File: docs/source/core/core.rst [hunk 4]
```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
...
+        # type1 will accept elements of type2 and type3 because it only cares about element kind (SpectrogramType)
+        # number of axes (3) and that first one corresponds to batch
+        assert type1.compare(type2) == NeuralTypeComparisonResult.SAME
+        assert type1.compare(type3) == NeuralTypeComparisonResult.INCOMPATIBLE
+
+
+
+Container types
+~~~~~~~~~~~~~~~
+NeMo type system understands Python containers (lists). So if your module returns a list of typed tensors, the way to express it by
```
Comment (titu1994) [inline docs/source/core/core.rst:867]: returns a "nested" list - we do support it. Dunno if any model actually uses more than depth 1 nest though so this sentence is fine too.

**Review by titu1994 — APPROVED** [review]
All looks good. Minor comments that would improve the docs a little.

File: docs/source/core/core.rst [hunk 5]
```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
...
 
     - Regular modules
     - Dataset/IterableDataset
     - Losses
 
+Every Neural Module in NeMo must inherit from `nemo.core.classes.module.NeuralModule` class.
+
+.. autoclass:: nemo.core.classes.module.NeuralModule
+
+Every Neural Modules inherits `nemo.core.classes.common.Typing` interface and needs to define neural types of it's inputs and outputs.
```
Comment (ericharper) [inline docs/source/core/core.rst:652]:
```suggestion
Every Neural Modules inherits the `nemo.core.classes.common.Typing` interface and needs to define neural types for its inputs and outputs.
```

File: docs/source/core/core.rst [hunk 6]
```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
...
+        )
+
+    @typecheck()
+    def forward(self, audio_signal, length=None):
+        ...
+
+The code snippet above means that `nemo.collections.asr.modules.conv_asr.ConvASREncoder` expects two arguments:
+    * First one, named `audio_signal` of shape [batch, dimension, time] with elements representing spectrogram values.
+    * Second one, named `length` of shape [batch] with elements representing lengths of corresponding signals.
+It also means that `.foraward(...)` and `__call__(...)` methods each produce two outputs:
```
Comment (ericharper) [inline docs/source/core/core.rst:684]:
```suggestion
It also means that `.forward(...)` and `__call__(...)` methods each produce two outputs:
```

File: docs/source/core/core.rst [hunk 7]
```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
...
+    * First one, of shape [batch, dimension, time] but with elements representing encoded representation (AcousticEncodedRepresentation class)
+    * Second one, of shape [batch], corresponding to their lengths.
+
+.. tip:: It is a good practice to define types and add `@typecheck()` decorator to your `.forward()` method once your module is ready for use by others.
+
+.. note:: The output of `.forward(...)` method will always be of type `torch.Tensor`, so it will work with any other Pytorch code. The type information will be carried with every tensor.
+If tensors without types will be passed to your module it will not fail but types will note be checked. Thus it is recommended to define input/output types for all your modules, starting with
```
Comment (ericharper) [inline docs/source/core/core.rst:691]:
```suggestion
If tensors without types will be passed to your module it will not fail but types will not be checked. Thus it is recommended to define input/output types for all your modules, starting with
```

File: docs/source/core/core.rst [hunk 8]
```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
...
+.. autoclass:: nemo.core.neural_types.NeuralTypeComparisonResult
+
+
+Examples
+--------
+
+Long vs short notation
+~~~~~~~~~~~~~~~~~~~~~~
+NeMo's NeuralType class allows you to express axis semantics information in long and short form.
+Consider these two equivalent types. Both encoder 3 dimensional tensors and both contain elements of type `AcousticEncodedRepresentation` (this type is a typical output of ASR encoders).
```
Comment (ericharper) [inline docs/source/core/core.rst:738]: Not sure how to fix but "Both encoder" seems wrong.

File: docs/source/core/core.rst [hunk 9]
```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
...
+
+.. code-block:: python
+
+    type1 = NeuralType(axes=('B', 'T', 'C'))
+    type2 = NeuralType(axes=('T', 'B', 'C'))
+    assert type1.compare(type2) == NeuralTypeComparisonResult.TRANSPOSE_SAME
+    assert type2.compare(type1) == NeuralTypeComparisonResult.TRANSPOSE_SAME
+
+Note that in this example, we dropped `elements_type` argument of `NeuralType` constructor.
+If note supplied, the element type would be VoidType.
```
Comment (ericharper) [inline docs/source/core/core.rst:764]:
```suggestion
If not supplied, the element type would be VoidType.
```

File: docs/source/core/core.rst [hunk 10]
```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
...
+        assert btc_void.compare(btc_spct_bad) == NeuralTypeComparisonResult.INCOMPATIBLE
+        # Note that even though VoidType is compatible with every other type, other types are not compatible with VoidType!
+        # It is one-way compatibility
+        assert btc_spctr.compare(btc_void) == NeuralTypeComparisonResult.INCOMPATIBLE
+
+Element type inheritance
+~~~~~~~~~~~~~~~~~~~~~~~~
+Neural types in NeMo support Python inheritance between element types.
+Consider an example where you want to develop a Neural Module which performs data augmentation for all kinds of spectrograms.
+In ASR two types of spectrograms are frequently used: mel and mfcc. To express this we will create two 3 classes to express
```
Comment (ericharper) [inline docs/source/core/core.rst:795]: Kind of hard to follow. What's meant by "two 3 classes"?

File: docs/source/core/core.rst (hunk 4 above)
Comment (ericharper) [inline docs/source/core/core.rst:867]:
```suggestion
NeMo type system understands Python containers (lists). If your module returns a list of typed tensors, the way to express it is by
```

File: docs/source/core/core.rst [hunk 11]
```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
...
+        # number of axes (3) and that first one corresponds to batch
+        assert type1.compare(type2) == NeuralTypeComparisonResult.SAME
+        assert type1.compare(type3) == NeuralTypeComparisonResult.INCOMPATIBLE
+
+
+
+Container types
+~~~~~~~~~~~~~~~
+NeMo type system understands Python containers (lists). So if your module returns a list of typed tensors, the way to express it by
+using Python list notation and Neural Types together like in the example below when defining your input/output types:
```
Comment (ericharper) [inline docs/source/core/core.rst:868]:
```suggestion
using Python list notation and Neural Types together when defining your input/output types:
```

**Review by ericharper — APPROVED** [review]
LGTM. Fixed some small typos. Left two comments asking for some more clarity.

File: docs/source/core/core.rst [hunk 12]
```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
...
+The code snippet above means that `nemo.collections.asr.modules.conv_asr.ConvASREncoder` expects two arguments:
+    * First one, named `audio_signal` of shape [batch, dimension, time] with elements representing spectrogram values.
+    * Second one, named `length` of shape [batch] with elements representing lengths of corresponding signals.
+It also means that `.foraward(...)` and `__call__(...)` methods each produce two outputs:
+    * First one, of shape [batch, dimension, time] but with elements representing encoded representation (AcousticEncodedRepresentation class)
+    * Second one, of shape [batch], corresponding to their lengths.
+
+.. tip:: It is a good practice to define types and add `@typecheck()` decorator to your `.forward()` method once your module is ready for use by others.
+
+.. note:: The output of `.forward(...)` method will always be of type `torch.Tensor`, so it will work with any other Pytorch code. The type information will be carried with every tensor.
```
[the remark this answers is not in the document]
Comment (okuchaiev) [reply docs/source/core/core.rst:690]: changed

File: docs/source/core/core.rst (hunk 1 above)
Comment (okuchaiev) [reply docs/source/core/core.rst:789]: no, this is intentionally non symmetric. Think of a function which accepts (void*)  --- it should accept every type. but (int) should only accept int and reject void.

File: docs/source/core/core.rst (hunk 2 above)
Comment (okuchaiev) [reply docs/source/core/core.rst:837]: yes

Comment (okuchaiev) [reply docs/source/core/core.rst:837]: I actually copied unittests :D

File: docs/source/core/core.rst (hunk 4 above)
Comment (okuchaiev) [reply docs/source/core/core.rst]: added

File: docs/source/core/core.rst [hunk 13]
```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
...
+                "outputs": NeuralType(('B', 'D', 'T'), AcousticEncodedRepresentation()),
+                "encoded_lengths": NeuralType(tuple('B'), LengthsType()),
+            }
+        )
+
+    @typecheck()
+    def forward(self, audio_signal, length=None):
+        ...
+
+The code snippet above means that `nemo.collections.asr.modules.conv_asr.ConvASREncoder` expects two arguments:
```
[the remark this answers is not in the document]
Comment (okuchaiev) [reply docs/source/core/core.rst:681]: done
